File: src/commands/moderation/kick.ts
/** @format */
//#region               External Libs
import { Ok } from "ts-results";
import {
    ChatInputCommandInteraction,
    Guild,
    PermissionFlagsBits,
    SlashCommandBuilder,
    User,
} from "discord.js";
//#endregion
//#region               Modules
import { types as commandTypes } from "@/system/handlers/command";
import {
    BotErr,
    ErrorKind,
    ErrorOrigin,
    Result,
} from "@/system/handlers/errHandlers";
import { PunishmentHandler } from "../../lib/system/handlers/punishmentHandler";
import { EmbedMessagesHandler } from "//lib/external/handlers/embed";
import { LOCALE } from "../..";
//#endregion
//#region               Typing
export namespace types {
    export type data = () => Promise<Ok<commandTypes.CommandData>>;
    export type action = (
        guild: Guild,
        user: User,
        moderator: User,
        createdAt: Date,
        reason: string
    ) => Promise<Ok<void>>;
    export type execute = (
        interaction: ChatInputCommandInteraction
    ) => Promise<Result<void>>;
}
//#endregion
//#region           Variables
const commandName = "kick";
//#endregion
//#region               Implementation
export const data: types.data = async () => {
    const description = {
        comm: "Aplique mute em alguém!",
        user: "Quem deseja mutar?",
        reason: "Qual o motivo do mute?",
        limit: "Por quanto tempo voce deseja manter essa pessoa mutada?",
        ephemeral: "Deseja que eu esconda essa mensagem?",
    };

    return new Ok({
        properties: {
            guild: false,
        },
        slashData: new SlashCommandBuilder()
            .setName(commandName)
            .setDescription(description.comm)
            .setDefaultMemberPermissions(PermissionFlagsBits.BanMembers)
            .addUserOption((user) =>
                user
                    .setName("user")
                    .setDescription(description.user)
                    .setRequired(true)
            )
            .addStringOption((reason) =>
                reason
                    .setName("reason")
                    .setDescription(description.reason)
                    .setRequired(true)
            )
            .addBooleanOption((ephemeral) =>
                ephemeral
                    .setName("ephemeral")
                    .setDescription(description.ephemeral)
                    .setRequired(false)
            )
            .setDMPermission(false),
    });
};
export const action: types.action = async (
    guild,
    user,
    moderator,
    createdAt,
    reason
) => {
    await guild.members.kick(user, reason);

    await new PunishmentHandler(guild).add(
        user.id,
        moderator.id,
        commandName,
        reason,
        createdAt,
        createdAt
    );

    return Ok.EMPTY;
};
export const execute: types.execute = async (interaction) => {
    const userToKick = interaction.options.getUser("user", true);
    const reason = interaction.options.getString("reason", true);

    const getEphemeral = interaction.options.getBoolean("ephemeral");
    await interaction.deferReply({
        ephemeral: getEphemeral !== null ? getEphemeral : true,
    });

    if (!interaction.guild)
        return new BotErr({
            message: "Guild not found!",
            kind: ErrorKind.NotFound,
            origin: ErrorOrigin.Unknown,
        });

    await action(
        interaction.guild,
        userToKick,
        interaction.user,
        interaction.createdAt,
        reason
    );

    const embed = (
        await new EmbedMessagesHandler("info.simpleResponse").Mount({
            title: "Warn",
            description:
                `O usuário ${userToKick} for removido do servidor com sucesso!\\n`,
            createdAt: interaction.createdAt.toLocaleString(LOCALE),
            username: `@${interaction.user.username}`,
        })
    ).unwrap();

    embed.setColor("#fc8000");

    await interaction.editReply({
        embeds: [embed.data],
    });

    return Ok.EMPTY;
};
//#endregion

File: src/commands/moderation/kick.rs
use poise::Context;
use serenity::all::{GuildId, User};

use crate::{
    handlers::{
        error::{BotErr, BotError, Error, ErrorKind, ErrorOrigin},
        punishment::PunishmentHandler,
    },
    prisma::PunishmentType,
    utils, BotData,
};

pub async fn action(
    ctx: Context<'_, BotData, BotErr>,
    guild_id: GuildId,
    user: User,
    reason: String,
) -> Result<(), BotErr> {
    let http = ctx.http();

    //let member = guild_id.member(&http, user.id).await?;
    //member.kick_with_reason(&http, reason.as_str()).await?;

    let create_punishment = PunishmentHandler {
        guild_id,
        prisma: ctx.data().prisma.clone(),
    }
    .add(
        user.id,
        ctx.author().id,
        PunishmentType::Kick,
        reason,
        ctx.created_at().to_utc(),
        ctx.created_at().to_utc(),
    )
    .await;

    if let Err(err) = create_punishment {
        return Err(BotError::new(
            err.to_string(),
            ErrorKind::Other,
            ErrorOrigin::Unknown,
        ));
    }

    Ok(())
}

/// Kick a user and specify a reason!
#[poise::command(slash_command)]
pub async fn kick(
    ctx: Context<'_, BotData, BotErr>,
    user: User,
    reason: String,
    ephemeral: Option<
        bool,
    >,
) -> Result<(), BotErr> {
    utils::matchs::ephemeral(&ctx, ephemeral).await?;

    let guild_id =
        utils::matchs::required_guild_id(&ctx, t!("use.command.only_server").to_string().as_str()).await?;

    action(ctx, guild_id, user, reason.to_string()).await?;

    ctx.say(t!("reply.kick.success")).await?;

    Ok(())
}
